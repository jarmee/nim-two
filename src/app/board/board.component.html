<button
  type="button"
  data-test-id="execute-play-one"
  (click)="onExecutePlay(1)"
>
  1
</button>
