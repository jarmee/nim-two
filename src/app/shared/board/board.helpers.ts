import {
  Board,
  BoardDifference,
  BoardDifferences,
  Column,
  Columns,
  Path
} from "./board.model";

function createBoardDifferencesOf(
  boardA: Board,
  andBoardB: Board
): BoardDifferences {
  const sourceBoard = boardA || andBoardB;
  return Object.keys(sourceBoard)
    .map(rowKey =>
      Object.keys(sourceBoard[rowKey]).map((columnKey: string) => ({
        newColumn: findColumnOf(boardA, [rowKey, columnKey]),
        currentColumn: findColumnOf(andBoardB, [rowKey, columnKey]),
        path: [rowKey, columnKey]
      }))
    )
    .reduce(
      (flatten: [Column, Column][], element: any) => [...flatten, ...element],
      []
    );
}

function columnsWithEqualValues(difference: BoardDifference): boolean {
  return (
    !difference.newColumn ||
    !difference.currentColumn ||
    difference.newColumn.value !== difference.currentColumn.value
  );
}

export function findColumnOf(board: Board, path: Path): Column | null {
  if (!board) return null;
  if (!path || !path.length) return null;

  return path.reduce((selectedProperty: any, pathElement: string) => {
    if (!selectedProperty || !selectedProperty[pathElement]) return null;
    return selectedProperty[pathElement];
  }, board);
}

export function differenceOf(
  boardA: Board,
  andBoardB: Board
): BoardDifferences {
  if (!boardA && !andBoardB) return [];

  return createBoardDifferencesOf(boardA, andBoardB).filter(
    columnsWithEqualValues
  );
}

export const withColumnValueFalseFilter = (column: Column) => !column.value;

export function countColumnsOf(
  board: Board,
  withFilter: (column: Column) => boolean
): number {
  if (!board) return 0;

  return Object.values(board)
    .map((columns: Columns) => Object.values(columns).filter(withFilter).length)
    .reduce(
      (amount: number, columnCount: number) => (amount += columnCount),
      0
    );
}
