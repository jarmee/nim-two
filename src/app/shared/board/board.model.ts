export type Column = {
  value: boolean;
  player?: string;
};
export type Columns = Record<number, boolean>;
export type Board = {
  [key: number]: Columns;
  errorMessage: string;
};
