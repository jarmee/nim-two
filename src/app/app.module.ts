import { BrowserModule } from "@angular/platform-browser";
import { NgModule } from "@angular/core";

import { AppComponent } from "./app.component";
import { BoardModule } from "./board/board.module";
import { GameEngineModule } from "./shared/game-engine/game-engine.module";

const initialGameState = {
  board: {
    0: {
      0: false,
      1: false,
      2: false,
      3: false,
      4: false,
      5: false,
      6: false,
      7: false,
      8: false,
      9: false,
      10: false,
      11: false,
      12: false
    }
  }
};

@NgModule({
  declarations: [AppComponent],
  imports: [
    BrowserModule,
    BoardModule,
    GameEngineModule.forRoot(initialGameState)
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule {}
