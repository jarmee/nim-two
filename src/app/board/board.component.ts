import { Component } from '@angular/core';
import { Observable } from 'rxjs';
import { GameEngineFacade } from '../shared/game-engine';
import { Board } from '../shared/game-engine/board/board.model';
import { GameStatus } from '../shared/game-engine/state/state.model';
import { Player } from '../shared/game-engine/turn/turn.model';

@Component({
  selector: 'app-board',
  templateUrl: './board.component.html',
  styleUrls: ['./board.component.scss']
})
export class BoardComponent {
  amount$: Observable<number> = this.gameEngine.amount$;

  board$: Observable<Board> = this.gameEngine.board$;

  status$: Observable<GameStatus> = this.gameEngine.status$;

  selectedPlayer$: Observable<Player> = this.gameEngine.selectedPlayer$;

  messages$: Observable<string> = this.gameEngine.messages$;

  constructor(private gameEngine: GameEngineFacade) {}

  onExecutePlay(board: Board) {
    this.gameEngine.executePlay(board);
  }

  onReset() {
    this.gameEngine.reset();
  }
}
