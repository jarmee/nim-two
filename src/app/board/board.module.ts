import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { BoardComponent } from "./board.component";
import { BoardFormComponent } from "./form/board-form.component";

@NgModule({
  declarations: [BoardComponent, BoardFormComponent],
  imports: [CommonModule],
  exports: [BoardComponent]
})
export class BoardModule {}
