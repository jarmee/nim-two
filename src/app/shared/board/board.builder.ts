import { Board } from "./board.model";

export class BoardBuilder {
  board: Board = {};

  addRowWithColumns(...columns: boolean[]): BoardBuilder {
    const currentIndex = Object.keys(this.board).length++;
    this.board = {
      ...this.board,
      [currentIndex]: {
        ...columns
          .map((value: boolean, index: number) => ({
            [index]: {
              value,
              player: null
            }
          }))
          .reduce((row: any, column: any) => ({ ...row, ...column }), {})
      }
    };
    return this;
  }

  build(): Board {
    return this.board;
  }

  static create(): BoardBuilder {
    return new BoardBuilder();
  }
}
