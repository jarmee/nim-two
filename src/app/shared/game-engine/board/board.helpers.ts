import { Board, BoardDifference, BoardDifferences, Column, Columns, Path } from './board.model';

function createBoardDifferencesOf(boardA: Board, andBoardB: Board): BoardDifferences {
  const sourceBoard = boardA || andBoardB;
  return Object.keys(sourceBoard)
    .map(rowKey =>
      Object.keys(sourceBoard[rowKey]).map((columnKey: string) => ({
        newColumn: findColumnOf(boardA, [rowKey, columnKey]),
        currentColumn: findColumnOf(andBoardB, [rowKey, columnKey]),
        path: [rowKey, columnKey]
      }))
    )
    .reduce((flatten: [Column, Column][], element: any) => [...flatten, ...element], []);
}

export function areColumnsDifferentByValue(newColumn: Column, currentColumn: Column): boolean {
  if (!newColumn && !currentColumn) {
    return false;
  }
  if (!newColumn || !currentColumn) {
    return true;
  }
  return newColumn.value !== currentColumn.value;
}

export function findColumnOf(board: Board, path: Path): Column | null {
  if (!board) {
    return null;
  }
  if (!path || !path.length) {
    return null;
  }

  return path.reduce((selectedProperty: any, pathElement: string) => {
    if (!selectedProperty || !selectedProperty[pathElement]) {
      return null;
    }
    return selectedProperty[pathElement];
  }, board);
}

export function differenceOf(boardA: Board, andBoardB: Board): BoardDifferences {
  if (!boardA && !andBoardB) {
    return [];
  }

  return createBoardDifferencesOf(boardA, andBoardB).filter(({ newColumn, currentColumn }: BoardDifference) =>
    areColumnsDifferentByValue(newColumn, currentColumn)
  );
}

export const withColumnValueFalseFilter = (column: Column) => !column.value;

export function countColumnsOf(board: Board, withFilter: (column: Column) => boolean): number {
  if (!board) {
    return 0;
  }

  return Object.values(board)
    .map((columns: Columns) => Object.values(columns).filter(withFilter).length)
    .reduce((amount: number, columnCount: number) => (amount += columnCount), 0);
}

export function exchangeColumnsOf(
  board: Board,
  exchangeColumnCallback: (column: Column, path?: Path) => Column
): Board {
  if (!board) {
    return null;
  }

  return Object.keys(board)
    .map(rowKey => ({
      [rowKey]: Object.keys(board[rowKey])
        .map(columnKey => ({
          [columnKey]: exchangeColumnCallback
            ? exchangeColumnCallback(board[rowKey][columnKey], [rowKey, columnKey])
            : { ...board[rowKey][columnKey] }
        }))
        .reduce((newRow, column) => ({ ...newRow, ...column }))
    }))
    .reduce((newBoard, row) => ({ ...newBoard, ...row }), {});
}
