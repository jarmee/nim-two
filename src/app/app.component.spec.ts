import { TestBed, async, ComponentFixture } from "@angular/core/testing";
import { AppComponent } from "./app.component";
import { BoardModule } from "./board/board.module";

describe("AppComponent", () => {
  let fixture: ComponentFixture<AppComponent>;
  let component: AppComponent;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [AppComponent],
      imports: [BoardModule]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(AppComponent);
    component = fixture.debugElement.componentInstance;
  });

  it("should create the app", () => {
    expect(component).toBeTruthy();
  });

  it("should match the snapshot", () => {
    expect(fixture).toMatchSnapshot();
  });
});
