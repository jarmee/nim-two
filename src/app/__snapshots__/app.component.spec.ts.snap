// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should match the snapshot 1`] = `
<app-root
  title={[Function String]}
>
  <app-board>
    <button
      data-test-id="execute-play-one"
      type="button"
    >
       1
  
    </button>
  </app-board>
</app-root>
`;
